crates/rpc/src/server/accept.rs

```diff
@@ -75,148 +182,232 @@ where
-            Ok(version) => version,
-            Err(e) => {
-                debug!(target: COMPONENT, "Failed to parse version: {}", e);
-                return bad_request("Invalid version specified in accept header value".into())
-                    .boxed();
-            },
-        };
+#[derive(Debug, thiserror::Error)]
+enum MediaRangeParsingError {
+    #[error("media range contained no types")]
+    MissingMediaType,
+    #[error("the media type {0} is invalid")]
+    InvalidMediaType(String),
+    #[error("the parameter {0} is invalid")]
+    InvalidParameter(String),
+    #[error("the quality weight value {0} is invalid")]
+    InvalidQuality(String),
+}
 
-        // Verify the version matches configured requirements
-        if self.version_req.matches(&version) {
-            self.inner.call(request).boxed()
-        } else {
-            debug!(target: COMPONENT, "Version does not match ({}/{})", version, self.version_req);
-            let msg = format!(
-                "Client / server version mismatch. Version {} is required by the server",
-                self.version_req
-            );
-            bad_request(msg).boxed()
-        }
+#[derive(Debug)]
+struct MediaRange<'a> {
+    main: MediaType<'a>,
+    subtype: MediaType<'a>,
+    params: HashMap<CaselessKey<'a>, &'a str>,
+    quality: Quality,
+}
+
+#[derive(Debug)]
+enum MediaType<'a> {
+    Wildcard,
+    Type(&'a str),
+}
+
+#[derive(Debug, Eq, Hash)]
+struct CaselessKey<'a>(&'a str);
+
+impl CaselessKey<'static> {
+    const VERSION: Self = Self("version");
+    const GENESIS: Self = Self("genesis");
+}
+
+impl PartialEq for CaselessKey<'_> {
+    fn eq(&self, other: &Self) -> bool {
+        self.0.eq_ignore_ascii_case(other.0)
     }
 }
 
-/// Returns a future that resolves to a bad request response.
-fn bad_request<B: Default + Send + 'static, E: Send + 'static>(
-    msg: String,
-) -> impl Future<Output = Result<http::Response<B>, E>> {
-    let response = http::Response::builder()
-        .status(http::StatusCode::BAD_REQUEST)
-        .header("content-type", "application/grpc")
-        .header("grpc-status", "3") // INVALID_ARGUMENT
-        .header("grpc-message", msg)
-        .body(B::default())
-        .expect("headers are valid");
+impl<'a> MediaType<'a> {
+    fn new(value: &'a str) -> Self {
+        match value.trim() {
+            "*" => Self::Wildcard,
+            other => Self::Type(other),
+        }
+    }
+
+    fn matches(&self, value: &str) -> bool {
+        match self {
+            MediaType::Wildcard => true,
+            MediaType::Type(media) => media.eq_ignore_ascii_case(value),
+        }
+    }
 
-    futures::future::ready(Ok(response))
+    fn starts_with(&self, prefix: &str) -> bool {
+        match self {
+            MediaType::Wildcard => false,
+            MediaType::Type(media) => {
+                media[..prefix.len().min(media.len())].eq_ignore_ascii_case(prefix)
```

Comment:
Would it be worth making `CaselessKey` transform the inner key into lowercase when instantiated? Then this comparison could read simpler (I think): `media.as_bytes().starts_with(prefix.as_bytes())`